### PR TITLE
Gati garch

### PR DESCRIPTION
**Pull request by GatiAher**
# Add GARCH-based dataset generator.

The GARCH dataset generator uses the ARCH package to fit GARCH models and generate synthetic data where the change in variance is time dependent, for example synthetic data with increasing or decreasing volatility. The default arguments initialize a GARCH(1,1) model, which is standard in the economic models literature. The GARCH model is fit to a univariate time series, in this case "Close" prices.

*Issue:* I don't know how to handle Volume. Right I just made it a series of 0.

---
Also, here are some pictures of the synthetic series. First 10 synthetic Close series generated by GARCH(1,1) with constant mean, fit to test data .csv.
Plot 10 GARCH(1,1) Generated Close Series

And 10 synthetic OHLC datasets generated by the GarchDatasetGenerator:

GARCH(1,1) Generated Data OHLC, series 1
GARCH(1,1) Generated Data OHLC, series 2
GARCH(1,1) Generated Data OHLC, series 3
GARCH(1,1) Generated Data OHLC, series 4
GARCH(1,1) Generated Data OHLC, series 5
GARCH(1,1) Generated Data OHLC, series 6
GARCH(1,1) Generated Data OHLC, series 7
GARCH(1,1) Generated Data OHLC, series 8
GARCH(1,1) Generated Data OHLC, series 9
GARCH(1,1) Generated Data OHLC, series 10

For comparison, plot of test data .csv
Test Data OHLC

---

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Pytest tests on transformations (e.g., transform series to percent change and back)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by GatiAher** [comment]
Issue: I don't know how to handle Volume. Right I just made it a series of 0.

File: src/coin_test/data/dataset_generator.py [hunk 1]
```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
...
+            "generalized error",
+        ] = "normal",
+        hold_back: int | None = None,
+        rescale: bool | None = None,
+    ) -> None:
+        """Initializes a GARCHDatasetGenerator."""
+        self.dataset = dataset
+        self.start: pd.Period = dataset.df.index[0]  # type: ignore
+        self.metadata = dataset.metadata
+        self.chunk_size = chunk_size
```
Comment (eito-fis) [inline src/coin_test/data/dataset_generator.py:292]: Chunk size doesn't seem to be used?

File: src/coin_test/data/dataset_generator.py [hunk 2]
```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
...
+        num_rows = len(pct_change_series) + 1
+        # Remove all series index before calculation.
+        pct_change_series = pct_change_series.reset_index(drop=True)
+
+        # Transform generated percent change values to price values.
+        synthetic_series = pd.Series(0, index=list(range(num_rows)))
+        # Intialize series_0 from starting value.
+        synthetic_series[0] = starting_price
+        # set series_j = series_j-1 + series_j-1 * pct_change_j-1 / 100.
+        for j in range(1, num_rows):
```
Comment (eito-fis) [inline src/coin_test/data/dataset_generator.py:468]: Would it be possible to do this with pandas operations? Maybe an apply over a rolling window or something? Not strictly necessary though.

**Review by eito-fis — CHANGES_REQUESTED** [review]
Looks really good! Left a couple comments, mostly small things.

File: src/coin_test/data/dataset_generator.py [hunk 3]
```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
...
+
+    Default values initialize a GARCH(1,1) model with constant mean.
+    """
+
+    mean: Literal[
+        "Constant", "Zero", "LS", "AR", "ARX", "HAR", "HARX", "constant", "zero"
+    ] = "Constant"
+    lags: int | NDArray | List[int] | None = 0
+    vol: Literal["GARCH", "ARCH", "EGARCH", "FIGARCH", "APARCH", "HARCH"] = "GARCH"
+    p: int | List[int] = 1
```
[the remark this answers is not in the document]
Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py]: Fixed

File: src/coin_test/data/dataset_generator.py (hunk 1 above)
Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py:292]: I use it when sampling chunks for defining High and Low series. The GARCH is only used to simulate Close series (which is also Open series when shifted by 1 period)

Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py:292]: See the sample_series function

File: src/coin_test/data/dataset_generator.py (hunk 2 above)
Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py:468]: I found this [stackoverflow post](https://stackoverflow.com/questions/34855859/is-there-a-way-in-pandas-to-use-previous-row-value-in-dataframe-apply-when-previ) for handling setting the value when the inputs to the calculation are not known in the beginning. I could use the apply function? Will get back to you on this.

File: src/coin_test/data/dataset_generator.py [hunk 4]
```diff
@@ -210,3 +214,396 @@ class ReturnsDatasetGenerator(StitchedChunkDatasetGenerator):
...
+        "gaussian",
+        "t",
+        "studentst",
+        "skewstudent",
+        "skewt",
+        "ged",
+        "generalized error",
+    ] = "normal"
+    hold_back: int | None = None
+    rescale: bool | None = None
```
[the remark this answers is not in the document]
Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py:246]: Good point. Fixed.

File: src/coin_test/data/dataset_generator.py (hunk 2 above)
Comment (eito-fis) [reply src/coin_test/data/dataset_generator.py:466]: Hmm, it does seem hard. Not a priority at all - I think its fine to leave as is.

**Review by eito-fis — APPROVED** [review]
Comments addressed, LGTM 👍

File: src/coin_test/data/dataset_generator.py (hunk 2 above)
Comment (GatiAher) [reply src/coin_test/data/dataset_generator.py:466]: From the stackoverflow post, the solution using apply also uses a global variable. Not standard but faster on average?
```python
import pandas as pd
import numpy as np

data = np.array([[10, 2, 10, 10],
                 [10, 3, 60, 100],
                 [np.nan] * 4,
                 [10, 22, 280, 250]]).T
idx = pd.date_range('20150131', end='20150203')
df = pd.DataFrame(data=data, columns=list('ABCD'), index=idx)
df
               A    B     C    D
 =================================
 2015-01-31    10   10    NaN  10
 2015-02-01    2    3     NaN  22 
 2015-02-02    10   60    NaN  280
 2015-02-03    10   100   NaN  250

def calculate(mul, add):
    global value
    value = value * mul + add
    return value

value = df.loc['2015-01-31', 'D']
df.loc['2015-01-31', 'C'] = value
df.loc['2015-02-01':, 'C'] = df.loc['2015-02-01':].apply(lambda row: calculate(*row[['A', 'B']]), axis=1)
df
               A    B     C     D
 =================================
 2015-01-31    10   10    10    10
 2015-02-01    2    3     23    22 
 2015-02-02    10   60    290   280
 2015-02-03    10   100   3000  250
```

**Pull request merged**